node-wasm/src/worker.rs

```diff
@@ -0,0 +1,372 @@
+use crate::worker::channel::{WorkerMessage, WorkerMessageServer};
+use crate::worker::commands::{NodeCommand, SingleHeaderQuery, WorkerResponse};
+use crate::wrapper::libp2p::NetworkInfoSnapshot;
+
+mod channel;
+pub(crate) mod commands;
+
+pub(crate) use channel::WorkerClient;
+
+const WORKER_MESSAGE_SERVER_INCOMING_QUEUE_LENGTH: usize = 64;
+
+/// actual type that's sent over a js message port
+type Result<T, E = WorkerError> = std::result::Result<T, E>;
+
+#[derive(Debug, Serialize, Deserialize, Error)]
+pub enum WorkerError {
+    #[error("node hasn't been started yet")]
+    NodeNotRunning,
+    #[error("node has already been started")]
+    NodeAlreadyRunning,
+    #[error("could not create blockstore: {0}")]
+    BlockstoreCrationFailed(String),
+    #[error("could not create header store: {0}")]
+    StoreCrationFailed(String),
+    #[error("could not parse genesis hash: {0}")]
+    GenesisHashInvalid(String),
+    #[error("could not parse bootstrap node multiaddr: {0}")]
+    BootstrapNodeMultiaddrInvalid(String),
+    #[error("node error: {0}")]
+    NodeError(String),
+    #[error("respose to command did not match expected type, should not happen")]
+    InvalidResponseType,
+    #[error("Value could not be serialized, should not happen: {0}")]
+    SerdeError(String),
+    #[error("command message could not be serialised, should not happen: {0}")]
+    CouldNotSerialiseCommand(String),
+    #[error("command response could not be serialised, should not happen: {0}")]
+    CouldNotSerialiseResponseValue(String),
+    #[error("response message could not be sent: {0}")]
+    CouldNotSendCommand(String),
+    #[error("Received empty worker response, should not happen")]
+    EmptyWorkerResponse,
+    #[error("Response channel to worker closed, should not happen")]
+    ResponseChannelDropped,
+}
+
+impl From<NodeError> for WorkerError {
+    fn from(error: NodeError) -> Self {
+        WorkerError::NodeError(error.to_string())
+    }
+}
+
+impl From<blockstore::Error> for WorkerError {
+    fn from(error: blockstore::Error) -> Self {
+        WorkerError::BlockstoreCrationFailed(error.to_string())
+    }
+}
+
+impl From<StoreError> for WorkerError {
+    fn from(error: StoreError) -> Self {
+        WorkerError::StoreCrationFailed(error.to_string())
+    }
+}
+
+impl From<TendermintError> for WorkerError {
+    fn from(error: TendermintError) -> Self {
+        WorkerError::GenesisHashInvalid(error.to_string())
+    }
+}
+
+impl From<MultiaddrError> for WorkerError {
+    fn from(error: MultiaddrError) -> Self {
+        WorkerError::BootstrapNodeMultiaddrInvalid(error.to_string())
+    }
+}
+
+impl From<serde_wasm_bindgen::Error> for WorkerError {
+    fn from(error: serde_wasm_bindgen::Error) -> Self {
+        WorkerError::SerdeError(error.to_string())
+    }
+}
+struct NodeWorker {
+    node: Node<IndexedDbStore>,
+}
+
+fn to_jsvalue_or_undefined<T: Serialize>(value: &T) -> JsValue {
+    to_value(value).unwrap()
+}
```

Comment:
I don't understand how `.unwrap()` achieves `undefined` there

Reply:
Right, missed it 👍